File: src/main/java/adventofcode/y20/D08.java
package adventofcode.y20;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import adventofcode.Day;
import adventofcode.InputReader;

public class D08 implements Day
{
	public Object star1() 
	{
		List<String> input = new InputReader(8,20).readLines();
		Set<Integer> exec = new HashSet<>();
		int ptr = 0;
		int acc = 0;
		
		while (!exec.contains(ptr))
		{
			String[] split = input.get(ptr).split(" ");
			exec.add(ptr);
			
			switch (split[0])
			{
				case "acc":
					acc += Integer.parseInt(split[1]);
					ptr++;
					break;
				case "jmp":
					ptr += Integer.parseInt(split[1]);
					break;
				default:
					ptr++;
			}
		}
		return acc;
	}

	public Object star2() 
	{
		List<String> input = new InputReader(8,20).readLines();
		int idx = 0;
		int ptr = 0;
		int acc = 0;
		
		do
		{
			List<String> cmds = new ArrayList<>(input);
			Set<Integer> exec = new HashSet<>();
			ptr = 0;
			acc = 0;
			
			// replace every jmp/nop command with its counterpart 
			while (cmds.get(idx).charAt(2) != 'p') idx++;
			if (cmds.get(idx).charAt(0)=='j') cmds.set(idx, "nop " + cmds.get(idx).substring(4));
			else cmds.set(idx,  "jmp " + cmds.get(idx).substring(4));
				
			// let program run over new code and check if it reaches the end
			while (!exec.contains(ptr) && 0<=ptr && ptr<cmds.size())
			{
				String[] split = cmds.get(ptr).split(" ");
				exec.add(ptr);
				
				switch (split[0])
				{
					case "acc":
						acc += Integer.parseInt(split[1]);
						ptr++;
						break;
					case "jmp":
						ptr += Integer.parseInt(split[1]);
						break;
					default:
						ptr++;
				}
			}
			idx++;
		}
		while (ptr<input.size());

		return acc;
	}
}

File: y20/d08.py
with open("y20/r08.txt", "r") as f:
    input = [x.strip().split(" ") for x in f.readlines()]

# star 1
visited = set()
acc = 0
pos = 0
while pos not in visited:
    visited.add(pos)
    if input[pos][0]=='acc':
        acc += int(input[pos][1])
        pos += 1
    elif input[pos][0]=='jmp':
        pos += int(input[pos][1])
    else:
        pos += 1
print(acc)

# star 2
for i in range(0, len(input)):
    copy = input.copy()
    if copy[i][0]=='jmp':
        copy[i] = ['nop', copy[i][1]]
    elif copy[i][0]=='nop':
        copy[i] = ['jmp', copy[i][1]]
    else:
        continue
    
    visited = set()
    acc = 0
    pos = 0
    while pos not in visited and pos!=len(input):
        visited.add(pos)
        if copy[pos][0]=='acc':
            acc += int(copy[pos][1])
            pos += 1
        elif copy[pos][0]=='jmp':
            pos += int(copy[pos][1])
        else:
            pos += 1
    
    if pos==len(input):
        print(acc)
        break

